docs/design/model-reader-writer-implementation-design.md

```diff
@@ -0,0 +1,244 @@
+    {
+        T Create(System.BinaryData data, System.ClientModel.Primitives.ModelReaderWriterOptions options);
+        string GetFormatFromOptions(System.ClientModel.Primitives.ModelReaderWriterOptions options);
+        System.BinaryData Write(System.ClientModel.Primitives.ModelReaderWriterOptions options);
+    }
+    public partial class ModelReaderWriterOptions
+    {
+        public ModelReaderWriterOptions(string format) { }
+        public string Format { get { throw null; } }
+        public static System.ClientModel.Primitives.ModelReaderWriterOptions Json { get { throw null; } }
+        public static System.ClientModel.Primitives.ModelReaderWriterOptions Xml { get { throw null; } }
+    }
+}
+```
+
+The methods in `IJsonModel<T>` controls how the model writes into a JSON or reads from JSON via `Utf8JsonWriter` or `Utf8JsonReader`. The methods in `IPersistableModel<T>` controls how the model writes into a BinaryData or reads from a BinaryData which serves a general purpose of serialization and deserialization including JSON case.
+
+### Implementation design
+
+Our general principals are:
+1. The behavior should keep unchanged when the library calls it on wire.
+2. When the customer calls the `ModelReaderWriter.Read` API, it should properly deserialize all the content in the input into the model (including the properties that this model does not currently support).
+3. When the customer calls the `ModelReaderWriter.Write` API, it should properly serialize all the properties into the result (including the properties that this model does not currently support).
+
+We should have the following methods in a model:
+
+1. `IUtf8JsonSerializable.Write(Utf8JsonWriter writer)` method.
+2. `IJsonModel<T>.Write(Utf8JsonWriter writer, ModelReaderWriterOptions options)` method.
+3. `IJsonModel<T>.Create(ref Utf8JsonReader reader, ModelReaderWriterOptions options)` method.
+4. `IPersistableModel<T>.Write(ModelReaderWriterOptions options)` method.
+5. `IPersistableModel<T>.Create(BinaryData data, ModelReaderWriterOptions options)` method.
+6. the static `Deserialize{ModelName}(JsonElement element, ModelReaderWriterOptions options = null)` method.
+
+If the model supports `xml` format, it will contain the following methods:
+1. `IXmlSerializable.Write(XmlWriter writer, string nameHint)` method.
+2. `WriteInternal(XmlWriter writer, string nameHint, ModelReaderWriterOptions options)` method.
+3. the static `Deserialize{ModelName}(XElement element, ModelReaderWriterOptions options = null)` method.
+
+#### The changes on models
+
+All models will now have a field `private IDictionary<string, BinaryData> _serializedAdditionalRawData`, if the model has a base type, the field should change to `private protected IDictionary<string, BinaryData> _serializedAdditionalRawData`.
+
+This private field works exactly the same as additional properties property, it should vanish when the model or any of its base type has an additional properties property defined.
```

Comment:
What if the base type has an additional properties defined as:
```
public IDictionary<string, string> AdditionalProperties { get; }  ?
```
Will we convert the type into  
```
private IDictionary<string, BinaryData>?
```

Reply:
this is indeed an issue. We need to resolve this later.
I will create an issue for this.

Reply:
Issue created: https://github.com/Azure/autorest.csharp/issues/4128